File: schemas/inference.py
"""Uploaded image records and capacity checks for experimental encode and decode."""

from typing import Annotated, Literal, Self

from pydantic import AwareDatetime, Field, model_validator

from schemas.base import StrictRecord
from schemas.images import ImageSummary
from schemas.models import ModelIdentifier
from schemas.protocol import PayloadCapacity

ImageReference = Annotated[str, Field(pattern=r"^image_[0-9a-f]{32}$")]
StoredReference = Annotated[str, Field(pattern=r"^(image|encoded)_[0-9a-f]{32}$")]
ImagePurpose = Literal["cover", "encoded"]
MAXIMUM_STORED_IMAGE_BYTES = 64 * 1024 * 1024
RETENTION_SECONDS = 24 * 60 * 60


class UploadedImage(StrictRecord):
    """Describe one accepted upload without its file name, path, or bytes."""

    image_reference: ImageReference
    purpose: ImagePurpose
    summary: ImageSummary
    warnings: list[str] = Field(default_factory=list, max_length=4)
    created_at: AwareDatetime
    expires_at: AwareDatetime

    @model_validator(mode="after")
    def validate_purpose_and_expiry(self) -> Self:
        """Tie the pixel policy to the purpose and keep the expiry after creation."""
        expected = "prepare_srgb" if self.purpose == "cover" else "preserve_stored"
        if self.summary.pixel_policy != expected:
            raise ValueError("The image summary does not match the upload purpose.")
        if self.expires_at <= self.created_at:
            raise ValueError("The image expiry must be after its creation time.")
        return self


class StoredImageRecord(UploadedImage):
    """Persist an upload or a published result with its stored PNG checksum."""

    image_reference: StoredReference
    stored_bytes: int = Field(ge=1, le=MAXIMUM_STORED_IMAGE_BYTES)
    checksum: str = Field(pattern=r"^[0-9a-f]{64}$")

    def public(self) -> UploadedImage:
        """Return the client-facing record without storage details."""
        return UploadedImage.model_validate(
            self.model_dump(exclude={"stored_bytes", "checksum"})
        )


class CapacityRequest(StrictRecord):
    """Ask how many message bytes one uploaded image can carry with one model."""

    image_reference: ImageReference
    model_identifier: ModelIdentifier


class CapacityResult(StrictRecord):
    """Report the message limit for one image and model inside the model range."""

    image_reference: ImageReference
    model_identifier: ModelIdentifier
    profile_identifier: Literal["test_only_v1"] = "test_only_v1"
    width: int = Field(ge=512, le=1024)
    height: int = Field(ge=512, le=1024)
    maximum_message_bytes: int = Field(ge=256, le=1024)
    capacity: PayloadCapacity

    @model_validator(mode="after")
    def validate_limit_matches_layout(self) -> Self:
        """Keep the advertised limit equal to the detailed layout accounting."""
        if self.maximum_message_bytes != self.capacity.maximum_message_bytes:
            raise ValueError("The message limit must match the layout capacity.")
        if self.capacity.payload_map_bits != self.width * self.height:
            raise ValueError("The layout capacity must match the image size.")
        return self

File: user_interface/src/contracts/inference.ts
/** Safe description of how an uploaded image was checked and prepared. */
export interface ImageSummary {
  source_format: 'JPEG' | 'PNG';
  mode: 'RGB' | 'RGBA';
  pixel_policy: 'prepare_srgb' | 'preserve_stored';
  source_width: number;
  source_height: number;
  prepared_width: number;
  prepared_height: number;
  input_size_bytes: number;
  orientation: number;
  color_policy: 'icc_to_srgb' | 'declared_srgb' | 'assumed_srgb';
}

/** One accepted upload, kept by the backend for at most one day. */
export interface UploadedImage {
  image_reference: string;
  purpose: 'cover' | 'encoded';
  summary: ImageSummary;
  warnings: string[];
  created_at: string;
  expires_at: string;
}

/** Ask how many message bytes one uploaded image carries with one model. */
export interface CapacityRequest {
  image_reference: string;
  model_identifier: string;
}

/** Layout accounting shared with the backend protocol, without model claims. */
export interface PayloadCapacity {
  profile_identifier: 'test_only_v1';
  maximum_message_bytes: number;
  header_bytes: number;
  message_length_bytes: number;
  authentication_bytes: number;
  padding_bytes_at_capacity: number;
  block_count: number;
  frame_bytes: number;
  correction_bytes: number;
  protected_bits: number;
  payload_map_bits: number;
  repeated_bits: number;
  minimum_repetitions: number;
  additional_repetitions: number;
}

/** The message limit for one image and one installed model. */
export interface CapacityResult {
  image_reference: string;
  model_identifier: string;
  profile_identifier: 'test_only_v1';
  width: number;
  height: number;
  maximum_message_bytes: number;
  capacity: PayloadCapacity;
}
